clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -680,35 +683,97 @@ namespace {
 QualType getItemType(const FieldDecl *FD) { return FD->getType(); }
 QualType getItemType(const CXXBaseSpecifier &BS) { return BS.getType(); }
 
+// These enable handler execution only when previous handlers succeed.
+template <typename... Tn>
+static bool handleField(FieldDecl *FD, QualType FDTy, Tn &&... tn) {
+  bool result = true;
+  std::initializer_list<int>{(result = result && tn(FD, FDTy), 0)...};
+  return result;
+}
+template <typename... Tn>
+static bool handleField(const CXXBaseSpecifier &BD, QualType BDTy,
+                        Tn &&... tn) {
+  bool result = true;
+  std::initializer_list<int>{(result = result && tn(BD, BDTy), 0)...};
+  return result;
+}
+
+template <typename T> struct bind_param { using type = T; };
+
+template <> struct bind_param<CXXBaseSpecifier &> {
+  using type = const CXXBaseSpecifier &;
+};
+
+template <> struct bind_param<FieldDecl *&> { using type = FieldDecl *; };
+
+template <> struct bind_param<FieldDecl *const &> { using type = FieldDecl *; };
+
+template <typename T> using bind_param_t = typename bind_param<T>::type;
+
+// This definition using std::bind is necessary because of a gcc 7.x bug.
+#define KF_FOR_EACH(FUNC, Item, Qt)                                            \
+  handleField(                                                                 \
+      Item, Qt,                                                                \
+      std::bind(static_cast<bool (std::decay_t<decltype(handlers)>::*)(        \
+                    bind_param_t<decltype(Item)>, QualType)>(                  \
+                    &std::decay_t<decltype(handlers)>::FUNC),                  \
+                std::ref(handlers), _1, _2)...)
+
+// The following simpler definition works with gcc 8.x and later.
+//#define KF_FOR_EACH(FUNC)                                                   \
+//  handleField(Field, FieldTy, ([&](FieldDecl *FD, QualType FDTy) {          \
+//                return handlers.f(FD, FDTy);                                \
+//              })...)
```

Comment:
Not sure that this comment is useful. @erichkeane , WDYT?

Reply:
I actually asked for it :)  When the day comes that the min-version of GCC is 8.x, I'd like us to have this note in place to act as a reminder that we can remove this hack.

Reply:
Okay.